### PR TITLE
docs :Expanded documentation coverage

### PR DESCRIPTION
**Pull request by DeleMike**
Added documentation for the `compiler`, `component parser`, and `file-based router` and updated the README to include Documentation links plus introducing a new section -  "What's New?!"

### CODE REVIEW

File: docs/file_based_router.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+### Introduction
+This document provides an overview of the file-based routing feature. This feature enables developers to implement navigation capabilities in their applications through the use of structured file-based routes.
+
+### Feature Overview
+The file-based routing feature in Starfyre allows developers to define and manage routes for their applications using individual files. Each file corresponds to a route, and the routing system automatically generates routes based on the filenames within a specified directory.
+
+## Implementation Details
+
+###  Pages Directory Structure
+To utilize the file-based routing feature, developers need to create a pages directory within their Starfyre application project. This directory serves as the location for storing the route-defining files.
```
Comment (sansyrox) [inline docs/file_based_router.md:12]:
```suggestion
To utilize the file-based routing feature, developers need to create a `pages` directory within their Starfyre application project. This directory serves as the location for storing the route-defining files.
```

File: docs/file_based_router.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    def populate_router(self):
+        # Method details...
+
+```
+
+### Route Generation
+When the FileRouter is initialized and the `populate_router` method is called, the router scans the files in the pages directory. Files with the `.fyre` extension are considered routes. The route names are derived from these filenames, with the .fyre extension removed and converted to lowercase. The router also handles special cases like avoiding the `__init__` route and checking for conflicts with the `index` route.
+
+### HTML Output and Main File Generation
+A function called `prepare_html_and_main` takes the list of generated route names and generates HTML output files for each route. Additionally, it creates the main execution file `(build/__main__.py)` that orchestrates rendering and writing HTML content to files. This main file imports and uses corresponding components for each route to render the HTML content.
```
Comment (sansyrox) [inline docs/file_based_router.md:33]: can we replace `orchestrates` with a simpler word?

File: docs/component_parser.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
+# Component Parser in Starfyre
+
+### Introduction
+This document provides an overview of the `ComponentParser` in Starfyre. This crucial feature ensures the correct parsing of PyML structure, validates tags, and maintains proper tag order in Starfyre components.
+
+### Feature Overview
+The `ComponentParser` is responsible for parsing PyML structures and ensuring that they adhere to specified rules. It plays a significant role in validating tags, their attributes, and their order within the component.
```
Comment (sansyrox) [inline docs/component_parser.md:7]:
```suggestion
The `ComponentParser` is responsible for parsing `PyML` structures and ensuring that they adhere to specified rules. It plays a significant role in validating tags, their attributes, and their order within the component.
```

File: docs/README.md [hunk 4]
```diff
@@ -77,15 +111,14 @@ Usage: python -m starfyre [OPTIONS]
...
-  --build BOOLEAN  Start the build package
+  --build BOOLEAN  Compile and build package
   --help           Show this message and exit.
 ```
 
 ## 🗒️ How to contribute
 
 ### 🏁 Get started
-Please read the code of conduct and go through CONTRIBUTING.md before contributing to Robyn. Feel free to open an issue for any clarifications or suggestions.
+Please read the code of conduct and go through `CONTRIBUTING.md` before contributing to Starfyre. Feel free to open an issue for any clarifications or suggestions.
```
Comment (sansyrox) [inline docs/README.md:121]: nicee loll

File: docs/README.md [hunk 5]
```diff
@@ -11,6 +13,38 @@ Starfyre is a library that allows you to build reactive frontends using only Pyt
...
+
+## 💥 What's new?! 
+### 🔥 Command Line Interface (CLI) Improvement
+
+With the recent enhancements to the Starfyre CLI, the process of building and compiling Starfyre projects has been streamlined and made more intuitive. The CLI commands have been refined to provide a simpler and more consistent experience for developers. Let's take a look at the changes that have been introduced:
+
+#### Previous Command (Before Improvement)
+```bash
+python3 starfyre --dev=True --path="test-application/" && python -m starfyre --build=True --path="test-application/"
+```
```
Comment (sansyrox) [inline docs/README.md:32]: we don't need this. We can delete this

File: docs/README.md [hunk 6]
```diff
@@ -11,6 +13,38 @@ Starfyre is a library that allows you to build reactive frontends using only Pyt
 
 Please note that Starfyre is still very naive and may be buggy, as it was developed in just five days. However, it is under active development and we welcome contributions to improve it. Whether you are a seasoned web developer or new to frontend development, we hope that you will find Starfyre to be a useful tool. Its intuitive API and simple, declarative style make it easy to get started, and its powerful features allow you to build sophisticated applications.
 
+## 📚 Documentation
+**[Compiler](compiler.md)**: Learn about the compiler and its functionality.
+
+**[ComponentParser](component_parser.md)**: Understand how the Component Parser works and its usage.
+
+**[File-Based Router](file_based_router.md)**: Explore the file-based routing feature and use it.
```
Comment (sansyrox) [inline docs/README.md:21]: do these links work when you open it as a website?

File: docs/README.md [hunk 7]
```diff
@@ -11,6 +13,38 @@ Starfyre is a library that allows you to build reactive frontends using only Pyt
...
+
+
+**🔥 File-Based Router**:
+```
+test-application
+├───pages
+    ├───home.fyre
+    ├───about.fyre
+```
+A new file-based routing feature has been introduced! Learn how to define routes using individual files in the pages directory [here](file_based_router.md).
```
Comment (sansyrox) [inline docs/README.md:47]: same question for this link

File: docs/component_parser.md [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+1) Create a PyML file for your Starfyre component.
+
+2) Define your component structure using PyML syntax.
+
+3) Use the ComponentParser class to parse the PyML content of your component.
+
+4) Handle any exceptions raised during parsing, including the UnknownTagError.
+
+5) Utilize the parsed component structure to generate dynamic content within your Starfyre application.
```
Comment (sansyrox) [inline docs/component_parser.md:52]: can we wrap all the Pyml to `PyML`

File: docs/compiler.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+## Exception Classes
+
+### InitFyreMissingError
+This exception class is raised when the `pages/__init__.fyre` file is missing in the project directory. It ensures that the `__init__.fyre` file is present to maintain the project's structure.
+
+### IndexFileConflictError
+Raised when a Fyre file named `index.fyre` is found in the pages folder. This conflicts with the code generation process, as `__init__.fyre` is transpiled into `index.html`. This exception prevents conflicts and ensures smooth code generation.
+
+## Transpiler Details
+The transpiler takes the intermediate representation (IR) of Fyre files and generates executable Python code. It accomplishes this through the following steps:
```
Comment (sansyrox) [inline docs/compiler.md:28]: can we also tell that `IR` are `Python` files

**Comment by sansyrox** [comment]
Hey @DeleMike 👋 

Thanks for the PR 😄  I have some suggestions.

Do let me know if you have any questions!

File: docs/file_based_router.md (hunk 1 above)
Comment (DeleMike) [reply docs/file_based_router.md:12]: Done

File: docs/component_parser.md (hunk 3 above)
Comment (DeleMike) [reply docs/component_parser.md:7]: Done

File: docs/README.md (hunk 5 above)
Comment (DeleMike) [reply docs/README.md:32]: I will, thanks!

File: docs/component_parser.md (hunk 8 above)
Comment (DeleMike) [reply docs/component_parser.md:52]: Done

File: docs/compiler.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
+# Compiler Implementation
+
+### Introduction
+This document outlines the role of the compiler in the Starfyre framework. The compiler is responsible for converting Fyre files into Python files with intermediate representations (IR), while the transpiler converts the IR into executable Python code.
+
+### Compiler Overview
+The compiler is a core component of Starfyre that facilitates the conversion of Fyre files into Python files. It performs the following key functions:
```
Comment (sansyrox) [inline docs/compiler.md:7]:
```suggestion
The compiler is a core component of Starfyre that facilitates the conversion of `.fyre` files into Python files. It performs the following key functions:
```

Comment (sansyrox) [inline docs/compiler.md:7]:
```suggestion
The compiler is a core component of Starfyre that facilitates the conversion of `.fyre` files into `.py` files. It performs the following key functions:
```

File: docs/compiler.md [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+- **Identifying Fyre Files**: The compiler scans the project directory to locate Fyre files that need to be compiled.
```
Comment (sansyrox) [inline docs/compiler.md:9]: same here -> Fyre -> `.fyre`

File: docs/compiler.md [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+
+- **Exception Handling**: The compiler introduces new exception classes to handle specific scenarios, such as the absence of `__init__.fyre` or conflicts due to the use of an `index.fyre` file.
+
+- **Parsing Fyre Files**: Each Fyre file is parsed to extract `Python`, `CSS`, `PyML`, `JS`, and client-side Python components. These components are then transpiled into intermediate representations.
```
Comment (sansyrox) [inline docs/compiler.md:13]: same here

File: docs/compiler.md [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+
+## Exception Classes
+
+### InitFyreMissingError
+This exception class is raised when the `pages/__init__.fyre` file is missing in the project directory. It ensures that the `__init__.fyre` file is present to maintain the project's structure.
+
+### IndexFileConflictError
+Raised when a Fyre file named `index.fyre` is found in the pages folder. This conflicts with the code generation process, as `__init__.fyre` is transpiled into `index.html`. This exception prevents conflicts and ensures smooth code generation.
+
+## Transpiler Details
```
Comment (sansyrox) [inline docs/compiler.md:27]: @DeleMike , should this be in a different file?

**Review by sansyrox — COMMENTED** [review]
Great work @DeleMike :smile: 

I have some minor suggestions

**Comment by DeleMike** [comment]
Hey @sansyrox, thanks for the feedback. I have made some changes to the PR.

**Review by sansyrox — APPROVED** [review]
LGTM! Great work @DeleMike 🔥

**Pull request merged**